### PR TITLE
add available device to test_epoch_metric.py #3335

### PR DESCRIPTION
**Pull request by BanzaiTokyo**

### CODE REVIEW

**Review by vfdev-5 — APPROVED** [review]

File: tests/ignite/metrics/test_epoch_metric.py [hunk 1]
```diff
@@ -51,73 +51,94 @@ def compute_fn(y_preds, y_targets):
...
     em.reset()
-    output1 = (torch.rand(4, 3), torch.randint(0, 2, size=(4, 3), dtype=torch.long))
+    output1 = (torch.rand(4, 3, device=device), torch.randint(0, 2, size=(4, 3), dtype=torch.long, device=device))
     em.update(output1)
-    output2 = (torch.rand(4, 3), torch.randint(0, 2, size=(4, 3), dtype=torch.long))
+
+    output2 = (torch.rand(4, 3, device=device), torch.randint(0, 2, size=(4, 3), dtype=torch.long, device=device))
     em.update(output2)
 
-    assert all([t.device.type == "cpu" for t in em._predictions + em._targets])
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_epoch_metric.py:66]: Let's keep this check and just adapt the device check

File: tests/ignite/metrics/test_epoch_metric.py [hunk 2]
```diff
@@ -51,73 +51,98 @@ def compute_fn(y_preds, y_targets):
...
         return 0.0
 
-    em = EpochMetric(compute_fn)
+    device = torch.device(available_device)
+    em = EpochMetric(compute_fn, device=device)
+    assert em._device == device
 
     em.reset()
-    output1 = (torch.rand(4, 3), torch.randint(0, 2, size=(4, 3), dtype=torch.long))
+    output1 = (torch.rand(4, 3, device=device), torch.randint(0, 2, size=(4, 3), dtype=torch.long, device=device))
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_epoch_metric.py:63]: You set both the metric input and metric internal accumulators to available device (i.e. `EpochMetric(compute_fn, device=device)`), but we first wanted to test input on cpu and metric's accumulator on available device.

Comment (BanzaiTokyo) [reply tests/ignite/metrics/test_epoch_metric.py:63]: do I understand it correctly that output1 and output2 are inputs and that in the first part of the test they should be on cpu?

Comment (vfdev-5) [reply tests/ignite/metrics/test_epoch_metric.py:63]: output1 and output2 should always be on cpu. yes, they are inputs to the metric and outputs from the model + ground truth.

Comment (BanzaiTokyo) [reply tests/ignite/metrics/test_epoch_metric.py]: I don't understand how you would compare tensors on different devices :(

File: tests/ignite/metrics/test_epoch_metric.py [hunk 3]
```diff
@@ -51,73 +51,96 @@ def compute_fn(y_preds, y_targets):
...
+    assert torch.equal(em._predictions[1].cpu(), output2[0].cpu())
+    assert torch.equal(em._targets[0].cpu(), output1[1].cpu())
+    assert torch.equal(em._targets[1].cpu(), output2[1].cpu())
     assert em.compute() == 0.0
 
     # test when y and y_pred are (batch_size, 1) that are squeezed to (batch_size, )
     em.reset()
     output1 = (torch.rand(4, 1), torch.randint(0, 2, size=(4, 1), dtype=torch.long))
     em.update(output1)
+
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_epoch_metric.py:79]: Remove this change

File: tests/ignite/metrics/test_epoch_metric.py [hunk 4]
```diff
@@ -51,73 +51,96 @@ def compute_fn(y_preds, y_targets):
...
     output2 = (torch.rand(4, 1), torch.randint(0, 2, size=(4, 1), dtype=torch.long))
     em.update(output2)
 
-    assert all([t.device.type == "cpu" for t in em._predictions + em._targets])
-    assert torch.equal(em._predictions[0], output1[0][:, 0])
-    assert torch.equal(em._predictions[1], output2[0][:, 0])
-    assert torch.equal(em._targets[0], output1[1][:, 0])
-    assert torch.equal(em._targets[1], output2[1][:, 0])
+    if available_device == "cpu":
+        assert all([t.device.type == "cpu" for t in em._predictions + em._targets])
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_epoch_metric.py:84]: Same here as https://github.com/pytorch/ignite/pull/3395/files#r2071285365

File: tests/ignite/metrics/test_epoch_metric.py [hunk 5]
```diff
@@ -51,73 +51,96 @@ def compute_fn(y_preds, y_targets):
...
+    output2 = (
+        torch.rand(4, 3, device=available_device),
+        torch.randint(0, 2, size=(4, 3), dtype=torch.long, device=available_device),
+    )
     em.update(output2)
-    output3 = (torch.rand(4, 3), torch.randint(0, 2, size=(4, 3), dtype=torch.long))
+    output3 = (
+        torch.rand(4, 3, device=available_device),
+        torch.randint(0, 2, size=(4, 3), dtype=torch.long, device=available_device),
+    )
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_epoch_metric.py:113]: Do not put outputX tensors to available_device

File: tests/ignite/metrics/test_epoch_metric.py [hunk 6]
```diff
@@ -51,73 +51,96 @@ def compute_fn(y_preds, y_targets):
...
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_epoch_metric.py:136]: Same here, keep output on cpu

File: tests/ignite/metrics/test_epoch_metric.py (hunk 3 above)
Comment (BanzaiTokyo) [reply tests/ignite/metrics/test_epoch_metric.py:79]: which one?

File: tests/ignite/metrics/test_epoch_metric.py [hunk 7]
```diff
@@ -188,19 +213,27 @@ def assert_data_fn(all_preds, all_targets):
...
+    output1 = (
+        torch.rand(4, 2, device=available_device),
+        torch.randint(0, 2, size=(4, 2), dtype=torch.long, device=available_device),
+    )
     em.update(output1)
-    output2 = (torch.rand(4, 2), torch.randint(0, 2, size=(4, 2), dtype=torch.long))
+    output2 = (
+        torch.rand(4, 2, device=available_device),
+        torch.randint(0, 2, size=(4, 2), dtype=torch.long, device=available_device),
+    )
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_epoch_metric.py:232]: Keep outputX on cpu

File: tests/ignite/metrics/test_epoch_metric.py [hunk 8]
```diff
@@ -188,19 +213,27 @@ def assert_data_fn(all_preds, all_targets):
...
     em.update(output2)
 
-    assert all([t.device.type == "cpu" for t in em._predictions + em._targets])
+    if available_device == "cpu":
+        assert all([t.device.type == "cpu" for t in em._predictions + em._targets])
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_epoch_metric.py:236]: Check device type for all available_device values

File: tests/ignite/metrics/test_epoch_metric.py (hunk 3 above)
Comment (vfdev-5) [reply tests/ignite/metrics/test_epoch_metric.py:79]:
```suggestion
```

File: tests/ignite/metrics/test_epoch_metric.py [hunk 9]
```diff
@@ -77,26 +78,28 @@ def compute_fn(y_preds, y_targets):
...
 
-def test_mse_epoch_metric():
+def test_mse_epoch_metric(available_device):
     def compute_fn(y_preds, y_targets):
         return torch.mean(((y_preds - y_targets.type_as(y_preds)) ** 2)).item()
 
-    em = EpochMetric(compute_fn)
+    em = EpochMetric(compute_fn, device=available_device)
+    assert em._device == torch.device(available_device)
+    dtype = torch.int32 if available_device == "mps" else torch.long
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_epoch_metric.py:95]: Is it really necessary? You do not do that in other tests like test_epoch_metric etc. Seems like It is not a hard requirement by MPS
